File: implementations/rust/src/crypto.rs
//! Cryptographic primitives for the Grith protocol.
//!
//! Provides Ed25519 signing/verification via `ed25519-dalek`, SHA-256 hashing
//! via `sha2`, JCS (RFC 8785) JSON canonicalization, and utility functions
//! for nonce generation, timestamps, and constant-time comparison.

use crate::GrithError;
use ed25519_dalek::{Signer, Verifier};
use rand::RngCore;
use sha2::{Digest, Sha256};

/// An Ed25519 key pair containing the signing key, verifying key, and hex-encoded public key.
pub struct KeyPair {
    pub signing_key: ed25519_dalek::SigningKey,
    pub verifying_key: ed25519_dalek::VerifyingKey,
    pub public_key_hex: String,
}

/// Generate a new Ed25519 key pair from cryptographically secure randomness.
///
/// Returns a `KeyPair` with a fresh 32-byte private key, the derived public key,
/// and the hex-encoded public key string.
pub fn generate_key_pair() -> Result<KeyPair, GrithError> {
    let mut rng = rand::thread_rng();
    let mut secret = [0u8; 32];
    rng.fill_bytes(&mut secret);
    let signing_key = ed25519_dalek::SigningKey::from_bytes(&secret);
    let verifying_key = signing_key.verifying_key();
    let public_key_hex = hex::encode(verifying_key.as_bytes());
    Ok(KeyPair {
        signing_key,
        verifying_key,
        public_key_hex,
    })
}

/// Reconstruct a `KeyPair` from a 32-byte private key.
///
/// # Errors
/// Returns `GrithError::CryptoError` if the byte slice is not exactly 32 bytes.
pub fn key_pair_from_private_key(bytes: &[u8]) -> Result<KeyPair, GrithError> {
    let secret: [u8; 32] = bytes
        .try_into()
        .map_err(|_| GrithError::CryptoError(format!("Private key must be 32 bytes, got {}", bytes.len())))?;
    let signing_key = ed25519_dalek::SigningKey::from_bytes(&secret);
    let verifying_key = signing_key.verifying_key();
    let public_key_hex = hex::encode(verifying_key.as_bytes());
    Ok(KeyPair {
        signing_key,
        verifying_key,
        public_key_hex,
    })
}

/// Sign a message with an Ed25519 signing key.
///
/// Returns the 64-byte signature as a `Vec<u8>`.
pub fn sign(message: &[u8], signing_key: &ed25519_dalek::SigningKey) -> Result<Vec<u8>, GrithError> {
    let signature = signing_key.sign(message);
    Ok(signature.to_bytes().to_vec())
}

/// Verify an Ed25519 signature against a message and verifying key.
///
/// Returns `true` if the signature is valid, `false` otherwise. Never panics
/// on malformed inputs -- any error is treated as an invalid signature.
pub fn verify(
    message: &[u8],
    signature: &[u8],
    verifying_key: &ed25519_dalek::VerifyingKey,
) -> bool {
    if signature.len() != 64 {
        return false;
    }
    let sig_bytes: [u8; 64] = match signature.try_into() {
        Ok(b) => b,
        Err(_) => return false,
    };
    let sig = ed25519_dalek::Signature::from_bytes(&sig_bytes);
    verifying_key.verify(message, &sig).is_ok()
}

/// Compute the SHA-256 hash of raw bytes and return it as a lowercase hex string.
pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    hex::encode(result)
}

/// Compute the SHA-256 hash of a UTF-8 string and return it as a lowercase hex string.
pub fn sha256_string(data: &str) -> String {
    sha256_hex(data.as_bytes())
}

/// Compute the SHA-256 hash of a JSON value after JCS canonicalization.
///
/// The value is first serialized via `canonicalize_json`, then hashed.
/// Two structurally equal objects always produce the same hash regardless
/// of key insertion order.
pub fn sha256_object(obj: &serde_json::Value) -> Result<String, GrithError> {
    let canonical = canonicalize_json(obj);
    Ok(sha256_string(&canonical))
}

/// Deterministic JSON serialization following JCS (RFC 8785).
///
/// Recursively sorts all object keys alphabetically before serializing.
/// Produces identical output regardless of key insertion order.
pub fn canonicalize_json(obj: &serde_json::Value) -> String {
    let sorted = sort_keys(obj);
    // serde_json::to_string produces compact JSON without extra whitespace
    serde_json::to_string(&sorted).unwrap_or_default()
}

/// Recursively sort all object keys in a JSON value.
fn sort_keys(value: &serde_json::Value) -> serde_json::Value {
    match value {
        serde_json::Value::Object(map) => {
            // Collect keys, sort, and rebuild the map
            let mut keys: Vec<&String> = map.keys().collect();
            keys.sort();
            let mut sorted_map = serde_json::Map::new();
            for key in keys {
                if let Some(v) = map.get(key) {
                    // Skip null values to match JS behavior where undefined values are omitted
                    sorted_map.insert(key.clone(), sort_keys(v));
                }
            }
            serde_json::Value::Object(sorted_map)
        }
        serde_json::Value::Array(arr) => {
            serde_json::Value::Array(arr.iter().map(sort_keys).collect())
        }
        other => other.clone(),
    }
}

/// Generate 32 random bytes for use as a cryptographic nonce.
pub fn generate_nonce() -> Vec<u8> {
    let mut rng = rand::thread_rng();
    let mut nonce = vec![0u8; 32];
    rng.fill_bytes(&mut nonce);
    nonce
}

/// Constant-time comparison of two byte slices.
///
/// Returns `true` only if both slices have the same length and identical contents.
/// The comparison time is proportional to the length of the slices, preventing
/// timing side-channel attacks.
pub fn constant_time_equal(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

/// Return the current UTC time as an ISO 8601 string.
///
/// Format: `YYYY-MM-DDTHH:MM:SS.sssZ`
pub fn timestamp() -> String {
    chrono::Utc::now().format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_and_sign_verify() {
        let kp = generate_key_pair().unwrap();
        let message = b"hello grith";
        let sig = sign(message, &kp.signing_key).unwrap();
        assert!(verify(message, &sig, &kp.verifying_key));
        assert!(!verify(b"tampered", &sig, &kp.verifying_key));
    }

    #[test]
    fn test_sha256() {
        let hash = sha256_string("hello");
        assert_eq!(hash, "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
    }

    #[test]
    fn test_canonicalize_json() {
        let obj: serde_json::Value = serde_json::json!({"z": 1, "a": 2});
        let canonical = canonicalize_json(&obj);
        assert_eq!(canonical, r#"{"a":2,"z":1}"#);
    }

    #[test]
    fn test_constant_time_equal() {
        assert!(constant_time_equal(b"abc", b"abc"));
        assert!(!constant_time_equal(b"abc", b"abd"));
        assert!(!constant_time_equal(b"ab", b"abc"));
    }

    #[test]
    fn test_key_pair_from_private_key() {
        let kp = generate_key_pair().unwrap();
        let bytes = kp.signing_key.to_bytes();
        let restored = key_pair_from_private_key(&bytes).unwrap();
        assert_eq!(kp.public_key_hex, restored.public_key_hex);
    }

    #[test]
    fn test_nonce_length() {
        let nonce = generate_nonce();
        assert_eq!(nonce.len(), 32);
    }
}

File: implementations/go/crypto.go
package nobulex

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
)

// KeyPair holds an Ed25519 key pair with a precomputed hex-encoded public key.
type KeyPair struct {
	PrivateKey   ed25519.PrivateKey
	PublicKey     ed25519.PublicKey
	PublicKeyHex string
}

// GenerateKeyPair generates a new Ed25519 key pair from cryptographically
// secure randomness.
func GenerateKeyPair() (*KeyPair, error) {
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("nobulex: failed to generate Ed25519 key pair: %w", err)
	}
	return &KeyPair{
		PrivateKey:   priv,
		PublicKey:     pub,
		PublicKeyHex: hex.EncodeToString(pub),
	}, nil
}

// KeyPairFromPrivateKey reconstructs a KeyPair from an existing Ed25519
// private key. The private key must be 64 bytes (Go's ed25519.PrivateKey
// format which includes the public key suffix).
func KeyPairFromPrivateKey(privateKey ed25519.PrivateKey) (*KeyPair, error) {
	if len(privateKey) != ed25519.PrivateKeySize {
		return nil, fmt.Errorf("nobulex: private key must be %d bytes, got %d", ed25519.PrivateKeySize, len(privateKey))
	}
	pub := privateKey.Public().(ed25519.PublicKey)
	keyCopy := make(ed25519.PrivateKey, len(privateKey))
	copy(keyCopy, privateKey)
	return &KeyPair{
		PrivateKey:   keyCopy,
		PublicKey:     pub,
		PublicKeyHex: hex.EncodeToString(pub),
	}, nil
}

// Sign signs message bytes with an Ed25519 private key and returns
// the 64-byte signature.
func Sign(message []byte, privateKey ed25519.PrivateKey) ([]byte, error) {
	if len(privateKey) != ed25519.PrivateKeySize {
		return nil, fmt.Errorf("nobulex: private key must be %d bytes, got %d", ed25519.PrivateKeySize, len(privateKey))
	}
	sig := ed25519.Sign(privateKey, message)
	return sig, nil
}

// Verify checks an Ed25519 signature against a message and public key.
// Returns false for any error (malformed key, truncated signature, etc.).
func Verify(message, signature []byte, publicKey ed25519.PublicKey) bool {
	if len(publicKey) != ed25519.PublicKeySize {
		return false
	}
	if len(signature) != ed25519.SignatureSize {
		return false
	}
	return ed25519.Verify(publicKey, message, signature)
}

// SHA256Hex computes the SHA-256 hash of data and returns it as a
// lowercase hex string.
func SHA256Hex(data []byte) string {
	h := sha256.Sum256(data)
	return hex.EncodeToString(h[:])
}

// SHA256String computes the SHA-256 hash of a UTF-8 string and returns
// it as a lowercase hex string.
func SHA256String(data string) string {
	return SHA256Hex([]byte(data))
}

// SHA256Object canonicalizes the object to JSON (RFC 8785) and then
// computes the SHA-256 hash, returning a lowercase hex string.
func SHA256Object(obj interface{}) (string, error) {
	canonical, err := CanonicalizeJSON(obj)
	if err != nil {
		return "", err
	}
	return SHA256String(canonical), nil
}

// CanonicalizeJSON produces a deterministic JSON serialization following
// JCS (RFC 8785). Object keys are sorted lexicographically at every
// nesting level. The output is identical regardless of the original
// key insertion order.
func CanonicalizeJSON(obj interface{}) (string, error) {
	sorted := sortKeys(obj)
	b, err := json.Marshal(sorted)
	if err != nil {
		return "", fmt.Errorf("nobulex: failed to marshal canonical JSON: %w", err)
	}
	return string(b), nil
}

// sortKeys recursively sorts map keys and processes all nested structures.
func sortKeys(value interface{}) interface{} {
	if value == nil {
		return nil
	}

	switch v := value.(type) {
	case map[string]interface{}:
		keys := make([]string, 0, len(v))
		for k := range v {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		ordered := &orderedMap{keys: keys, values: make(map[string]interface{}, len(v))}
		for _, k := range keys {
			child := v[k]
			if child == nil {
				// Preserve explicit null values
				ordered.values[k] = nil
				ordered.hasNil = append(ordered.hasNil, k)
			} else {
				ordered.values[k] = sortKeys(child)
			}
		}
		return ordered
	case []interface{}:
		result := make([]interface{}, len(v))
		for i, item := range v {
			result[i] = sortKeys(item)
		}
		return result
	default:
		return value
	}
}

// orderedMap preserves key order during JSON marshaling.
type orderedMap struct {
	keys   []string
	values map[string]interface{}
	hasNil []string // keys that have explicit nil values
}

func (o *orderedMap) MarshalJSON() ([]byte, error) {
	var buf strings.Builder
	buf.WriteByte('{')
	nilSet := make(map[string]bool, len(o.hasNil))
	for _, k := range o.hasNil {
		nilSet[k] = true
	}
	for i, k := range o.keys {
		if i > 0 {
			buf.WriteByte(',')
		}
		keyBytes, err := json.Marshal(k)
		if err != nil {
			return nil, err
		}
		buf.Write(keyBytes)
		buf.WriteByte(':')
		v := o.values[k]
		if v == nil && nilSet[k] {
			buf.WriteString("null")
		} else if v == nil {
			buf.WriteString("null")
		} else {
			valBytes, err := json.Marshal(v)
			if err != nil {
				return nil, err
			}
			buf.Write(valBytes)
		}
	}
	buf.WriteByte('}')
	return []byte(buf.String()), nil
}

// ToHex encodes a byte slice to a lowercase hex string.
func ToHex(data []byte) string {
	return hex.EncodeToString(data)
}

// FromHex decodes a hex string to a byte slice.
func FromHex(hexStr string) ([]byte, error) {
	b, err := hex.DecodeString(hexStr)
	if err != nil {
		return nil, fmt.Errorf("nobulex: invalid hex string: %w", err)
	}
	return b, nil
}

// GenerateNonce generates 32 cryptographically secure random bytes.
func GenerateNonce() ([]byte, error) {
	nonce := make([]byte, 32)
	_, err := rand.Read(nonce)
	if err != nil {
		return nil, fmt.Errorf("nobulex: failed to generate nonce: %w", err)
	}
	return nonce, nil
}

// ConstantTimeEqual compares two byte slices in constant time to
// prevent timing side-channel attacks.
func ConstantTimeEqual(a, b []byte) bool {
	return subtle.ConstantTimeCompare(a, b) == 1
}

// Timestamp returns the current time as an ISO 8601 UTC string
// (e.g. "2025-01-15T12:00:00.000Z").
func Timestamp() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
}

// objectToMap converts any Go value to a map[string]interface{} via
// JSON round-trip. This is used internally to canonicalize arbitrary
// struct types.
func objectToMap(obj interface{}) (map[string]interface{}, error) {
	b, err := json.Marshal(obj)
	if err != nil {
		return nil, err
	}
	var m map[string]interface{}
	if err := json.Unmarshal(b, &m); err != nil {
		return nil, err
	}
	return m, nil
}
